### PR TITLE
disco_l475_iot1: add openocd support

### PR DESCRIPTION
**Pull request by mbolivar**
This requires openocd version 0.10 or later.

### CODE REVIEW

**Comment by mbolivar** [comment]
@erwango FYI. The openocd.cfg I got from Christophe.

**Review by galak — CHANGES_REQUESTED** [review]
Can you look at maybe moving OPENOCD_LOAD_CMD & OPENOCD_VERIFY_CMD into Makefile.inc/Makefile and setting up a default, something like:

ifeq ($(ARCH),arm)
OPENOCD_LOAD_CMD ?= "flash write_image erase ${O}/${KERNEL_BIN_NAME} ${CONFIG_FLASH_BASE_ADDRESS}"
OPENOCD_VERIFY_CMD ?= "verify_image ${O}/${KERNEL_BIN_NAME} ${CONFIG_FLASH_BASE_ADDRESS}"

export OPENOCD_LOAD_CMD OPENOCD_VERIFY_CMD
endif

**Comment by nashif** [comment]
> Can you look at maybe moving OPENOCD_LOAD_CMD & OPENOCD_VERIFY_CMD into Makefile.inc/Makefile and setting up a default, something like:

Maybe in the openocd.sh script, not the build system or top level Makefile...

**Comment by galak** [comment]
I'm not sure how we would do this in the script as its the Makefile(s) that are using the variables.

**Comment by mbolivar** [comment]
I've actually already started work on rewriting flash support via a Python script. It's nearly ready to submit as a PR; here is a WIP branch:

https://github.com/mbolivar/zephyr/tree/python-flash-scripts

This reimplements the shell scripts in Python, but compatibly using the same environment variables. That gives us a gradual transition path for the script to be converted to take command line arguments instead, so things like OPENOCD_LOAD_CMD etc. could go away entirely.

(After that I plan on adding debugging support and making the scripts DT-aware like the flashing scripts we have out of tree.)

Since this environment variable strategy is used by all other boards, would it be OK to take this patch as tech debt for now and let me clean it up later? I'm relying on this as a test for replacing openocd.sh in the Python script.

File: boards/arm/disco_l475_iot1/support/openocd.cfg [hunk 1]
```diff
@@ -0,0 +1,7 @@
+source [find interface/stlink-v2-1.cfg]
+
+transport select hla_swd
+
+source [find target/stm32l4x.cfg]
+
+reset_config srst_only
```
Comment (galak) [inline boards/arm/disco_l475_iot1/support/openocd.cfg]: For most of the openocd.cfg we have, curious if we dont need it:

$_TARGETNAME configure -event gdb-attach {
        echo "Debugger attaching: halting execution"
        reset halt
        gdb_breakpoint_override hard
}

$_TARGETNAME configure -event gdb-detach {
        echo "Debugger detaching: resuming execution"
        resume
}

**Comment by galak** [comment]
I have no issues with having the cleanup as tech debt

File: boards/arm/disco_l475_iot1/support/openocd.cfg (hunk 1 above)
Comment (erwango) [reply boards/arm/disco_l475_iot1/support/openocd.cfg]: Actually, I don't even know how this is called/used, "make debug" does not seem to require those.

**Review by erwango — APPROVED** [review]
Sounds OK.
Though I'll raise an issue to ask for openocd 0.10.0 in SDK, so eveyone could use it

File: boards/arm/disco_l475_iot1/support/openocd.cfg (hunk 1 above)
Comment (mbolivar) [reply boards/arm/disco_l475_iot1/support/openocd.cfg]: I'm not sure -- I obtained this file from Christophe directly. Perhaps Erwan can comment.

**Comment by galak** [comment]
@erwango the 0.9.2 sdk should have openocd 0.10.0

**Comment by nashif** [comment]
> I'm not sure how we would do this in the script as its the Makefile(s) that are using the variables.

The variables are exported in the Makefile so they can be used by `scripts/support/openocd.sh`, so the check can be done there.

**Comment by mbolivar** [comment]
> Sounds OK.

Thanks, @erwango -- and does the .cfg need any changes as asked by @galak ?

**Comment by erwango** [comment]
@martib: I had a test with sdk 0.9.2 and make flash/make debug works nicely as is.
Regarding gdb-attach/gdb-detach, I actually don't know how this should be used.

So I guess we can have a try with your proposal, and we'll see if someone complains

**Review by nashif — APPROVED** [review]

**Comment by mbolivar** [comment]
> So I guess we can have a try with your proposal, and we'll see if someone complains

Thanks.

I'm moving on to doing debugging in Python, so I'll be sure to run into any issues on this board if this config is not working with that.

**Pull request merged**